api/client.go

```diff
@@ -728,16 +730,31 @@ func (c *client) RefsDump(ctx context.Context, repository string) (*models.RefsD
 	return resp.GetPayload(), nil
 }
 
-func NewClient(endpointURL, accessKeyID, secretAccessKey string) (Client, error) {
+type ClientOption func(*client)
+
+func MaxIdleConnsPerHost(maxIdelConnsPerHost int) ClientOption {
+	return func(c *client) {
+		c.transport.Transport.(*http.Transport).MaxIdleConnsPerHost = maxIdelConnsPerHost
+	}
+}
+
+func NewClient(endpointURL, accessKeyID, secretAccessKey string, opts ...ClientOption) (Client, error) {
 	parsedURL, err := url.Parse(endpointURL)
 	if err != nil {
 		return nil, err
 	}
 	if len(parsedURL.Path) == 0 {
 		parsedURL.Path = path.Join(parsedURL.Path, genclient.DefaultBasePath)
 	}
-	return &client{
-		remote: genclient.New(httptransport.New(parsedURL.Host, parsedURL.Path, []string{parsedURL.Scheme}), strfmt.Default),
-		auth:   httptransport.BasicAuth(accessKeyID, secretAccessKey),
-	}, nil
+	transport := httptransport.New(parsedURL.Host, parsedURL.Path, []string{parsedURL.Scheme})
+	transport.Transport.(*http.Transport).MaxIdleConnsPerHost = 100 // make this configurable?
+	c := &client{
+		transport: transport,
+		remote:    genclient.New(transport, strfmt.Default),
+		auth:      httptransport.BasicAuth(accessKeyID, secretAccessKey),
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
```

Comment:
OpenAPI uses the defaults from http.DefaultTransport and gives you the option to replace the client.
I'm worried that if we keep this path and try to configure every aspect in the client using options it will not end well or break in the case for example we will want to use some of the options the generated code enables us.

```go
	clientTransport := httptransport.NewWithClient(parsedURL.Host, parsedURL.Path, []string{parsedURL.Scheme}, client)
	clientTransport.EnableConnectionReuse()
	clientTransport.DefaultAuthentication = httptransport.BasicAuth(accessKeyID, secretAccessKey)
```
if we `EnableConnectionReuse` for example the implementation replaces the underlying Transport and the cast we introduced will break.

I suggest we configure the client we want and the default authentication.
We also need to consider using the generated client directly as for now it doesn't look like the only value is the option to have a flat interface and not a fluent API like swagger generates.

Reply:
not having to pass authentication info on every call but rather once on initialization is the main reason the client wrapper was introduced, along with encapsulating the domain knowledge of how to assemble a Swagger generated client

It is encapsulated by desgin - the constructor wraps the logic of how to assemble a swagger generated client, so we can rely on casting because we know the type of transport is only controlled by us.